### DISCUSSION TITLE
Backup MX with Postfixadmin

### DISCUSSION THREAD

**Discussion by fongaboo** [postfixadmin/postfixadmin#713] [Q&A]
I have my primary mail server running postfix with MySQL backend and PostfixAdmin front-end. Works as advertised.

For years my backup MX has just been configured to accept all mail for configured domains, and store-and-forward. But we know that hasn't been a good practice for a long time.

I have finally set up my backup MX with the same postfix/MySQL/postfixadmin stack as the primary. I see you can check off Backup MX as an option in Domains.

But I want to populate the backup MX with valid users so that anything else gets denied. How do you do that in a Postfixadmin setting?

Do aliases have to be created pointing to the same destination as is set on the primary?

How are addresses that are actual POP/IMAP mailboxes on the primary handled? Do you have to make an actual mailbox on the secondary? Or can you just make a matching alias? If the latter, what to point the aliases to?

Should I be planning to migrate the entire MySQL database as-is to the secondary server, but then have some script that sets backupmx=1?

TIA

**Comment by fongaboo** [comment]
So I did write a script to export domains, aliases and mailboxes to .sql files and then import them into the postfix database on the secondary.

If an address was a mailbox on the primary, the script just creates an alias on the secondary that points to postmaster@secondaryMX.domain.tld. But where the alias is pointing to on the secondary shouldn't matter, no? Because the domain is set to backupmx=1. So I would think it would just ignore where it is set to point to on the secondary, and instead forward it to the primary and then the settings on the primary should be followed.

telnet'ing to port 25 from an outside server:

**$ telnet secondaryMX.domain.tld 25
Trying <Secondary MX IP address>...
Connected to secondaryMX.domain.tld.
Escape character is '^]'.
220 secondaryMX.domain.tld ESMTP Postfix (Debian/GNU)
helo mydomain.tld
250 secondaryMX.domain.tld
mail from: <outside email>
250 2.1.0 Ok
rcpt to: alias@mydomain.tld
250 2.1.5 Ok
data
354 End data with <CR><LF>.<CR><LF>
this is being sent to the secondary
secondary accepted
the address is an alias on the secondary
but it is an actual mailbox on the primary
the domain is set to backupmx=1 on the secondary
what will it do? where will it go?
.
250 2.0.0 Ok: queued as 755313A3135**


**fgrep 755313A3135 /var/log/mail.log
Mar  1 09:34:35 postfix/smtpd[2788220]: 755313A3135: client=<IP of outside server>
Mar  1 09:35:20 postfix/cleanup[2789496]: 755313A3135: message-id=<>
Mar  1 09:35:20 postfix/qmgr[235804]: 755313A3135: from=<outside email>, size=447, nrcpt=1 (queue active)
Mar  1 09:35:20 postfix/local[2790456]: 755313A3135: to=<root@secondaryMX.domain.tld>, orig_to=<alias@mydomain.tld>, relay=local, delay=52, delays=52/0.01/0/0.04, dsn=2.0.0, status=sent (delivered to mailbox)
Mar  1 09:35:20 postfix/qmgr[235804]: 755313A3135: removed**


Nothing arrives on the primary MX. But look what I find in the postmaster mailbox on the secondary:

**From <outside email>  Wed Mar  1 09:35:20 2023
Return-Path: <outside email>
X-Original-To: alias@mydomain.tld
Delivered-To: postmaster@secondaryMX.domain.tld
Received: from mydomain.tld (<outside server IP address>)
        by secondaryMX.domain.tld (Postfix) with SMTP id 755313A3135
        for <alias@mydomain.tld>; Wed,  1 Mar 2023 09:34:28 -0500 (EST)**

**this is being sent to the secondary
secondary accepted
the address is an alias on the secondary
but it is an actual mailbox on the primary
the domain is set to backupmx=1 on the secondary
what will it do? where will it go?**


mydomain.tld is set to backupmx=1 in Postfixadmin, yet it still delivers locally.

**Reply by fongaboo** [reply]
> Date: Wed, 1 Mar 2023 03:03:39
> From: Michael Neuffer <neuffer@neuffer.com>
> To: postfixadmin-devel@lists.sourceforge.net
> Subject: Re: [Postfixadmin-devel] Backup MX with Postfixadmin
> 
> @Janfred I assume both your mail servers are accessing the same DB backend?
> or are you replicating the information somehow  from the primary to the
> DB on the second mail server?

Not using the same database backend. I have a script on the primary that outputs .sql files. It outputs the aliases as-is. But it modifies the domains so that backupmx=1. Then it modifies mailboxes to be aliases.

**Reply by Janfred** [reply]
My Mailservers have a mariadb master-master-replication, so they basically run on the same but replicated database.

**Reply by Janfred** [reply]
AFAIR used the flag `backupmx` to identify a domain with a catch-all, so that for these domains mails are accepted regardless of an existing alias.

**Reply by fongaboo** [reply]
> My Mailservers have a mariadb master-master-replication, so they basically run on the same but replicated database.

Does your secondary store-and-forward to the primary?

**Reply by Janfred** [reply]
Yes, exactly. Except for forwards or Domains where my server is just the "proxy".
Due to silly firewall configuration where Port 25 is blocked my servers were the "public facing" mail servers for some domains and they just forwarded every mail.
This forward and the resolution of forwards with external targets were done directly by the secondary.
All mails that were intended to be delivered locally are store-and-forward.

**Reply by Janfred** [reply]
(I just realized, maybe there still is a bug in there, that forwards are executed twice if there is a forward+deliver to local mailbox configured. Need to check that.)
